### ISSUE TITLE
IO Error on installing HTTPFS extension for v0.9.1 CLI

### ISSUE THREAD

**Issue by ketozhang** [duckdb/duckdb#9378]
### What happens?

HTTPFS extension fails to install on duckdb CLI v0.9.1. The issue does not exist in the main branch with `http://extensions.duckdb.org/7127284469/osx_amd64/HTTPFS.duckdb_extension.gz`

### To Reproduce

```console
$ brew install duckdb
$ duckdb
```
```sql
D INSTALL HTTPFS;
Error: HTTP Error: Failed to download extension "httpfs" at URL "http://extensions.duckdb.org/v0.9.1/osx_amd64/httpfs.duckdb_extension.gz"
Extension "httpfs" is an existing extension.
```

### OS:

MacOS Sonoma 14.0

### DuckDB Version:

v0.9.1 401c8061c6

### DuckDB Client:

CLI

### Full Name:

Keto Zhang

### Affiliation:

Caltech

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by carlopi** [comment]
Can you try a few times?

I was testing this (since we touched the deployment of the httpfs extensions), I think there is some transient issues specific to this extension.

**Comment by ketozhang** [comment]
Attempted it a few times—no luck in the past hour :/

**Comment by ketozhang** [comment]
Yep that works.

**Comment by carlopi** [comment]
Thanks for reporting this, it's not the regular experience. Backstory is that we updated an extension, I think for the first time ever to solve a serious problem, and we are just discovering some pain points.

We should probably consider a secondary deployment to potentially point people to, or some other strategy to reduce impact of this in future similar situations.

**Comment by ketozhang** [comment]
No worries, thank you for the instant responses so far!

In the meantime, the nightly extension or downgrading to v0.9.0 or v0.8.1 is working for me.

**Comment by whscullin** [comment]
Is there a plan to have it be uploaded to the standard place, or do I need to update my code to point to the nightlies in order to use 0.9.1 ?

**Comment by whscullin** [comment]
It appears that the x86 darwin binary does not exist at the nightly location.

**Comment by carlopi** [comment]
@whscullin: Are you also affected?

Idea is that using regular extension repository should "just work".

IFF you or your user are experiencing any issue with the download of httpfs for v0.9.1, or if you want to mitigate this transitory situation, you can add a `INSTALL httpfs FROM 'http://nightly-extensions.duckdb.org';`. at the top of the script.

Note that this is relevant only for `v0.9.1` and only for `httpfs` extension.

**Comment by mustafahasankhan** [comment]
CALL load_aws_extension() is also failing due to the same.
Missing Extension Error: httpfs extension is required for load_aws_credentials

https://extensions.duckdb.org/v0.9.1/osx_arm64/httpsfs.duckdb_extension.gz is unreachable, shows Access Denied.

**Comment by carlopi** [comment]
@mustafahasankhan, or anyone else affected, can you do: `FORCE INSTALL httpfs FROM 'http://nightly-extensions.duckdb.org';`

This installs from the secondary official repository a signed httpfs extension.
This needs to be performed only once, then the asset is saved in your duckdb extension folder (defaults to `~/.duckdb/extensions/v0.9.1/<platform>/httpfs.duckdb_extension`).

**Comment by carlopi** [comment]
We will fix this (long) but transient issue this morning, I will update when state should be back to normal.

**Comment by mustafahasankhan** [comment]
@carlopi 
```
FORCE INSTALL httpfs FROM 'http://nightly-extensions.duckdb.org';
LOAD httpfs
```

gives:
IO Error: Extension "/Users/mustafa.khan/.duckdb/extensions/v0.9.1/osx_arm64/httpsfs.duckdb_extension" not found.

**Comment by Mause** [comment]
> @carlopi 
> ```
> FORCE INSTALL httpfs FROM 'http://nightly-extensions.duckdb.org';
> LOAD httpfs
> ```
> 
> gives:
> IO Error: Extension "/Users/mustafa.khan/.duckdb/extensions/v0.9.1/osx_arm64/httpsfs.duckdb_extension" not found.

Can you double check the name you tried? The error mentions httpsfs, which is incorrect

**Comment by carlopi** [comment]
We properly cleared CDN caches, and conducted some more testing, the proper `httpfs.duckdb_extension` should be available doing `INSTALL httpfs` (this is a no-op if you already have a httpfs.duckdb_extension locally) or `FORCE INSTALL httpfs` (that would download and potentially override any previous one).

You can test this by doing `FORCE INSTALL httpfs`, that would force a download and fail if download fails.

I trust this to be now solved, I will leave this open another few hours if anyone has feedback.

Thanks a lot everyone.

**Comment by begelundmuller** [comment]
Hello @carlopi, we had a user see this same issue today – they reported the issue in the Rill Discord [here](https://discord.com/channels/959796421832343562/959801632151969792/1166108814676856964)

**Comment by carlopi** [comment]
I think I miss some context, and the Discord link is not public available.

Could you open an new issue with the relevant details (even just the error message should be enough, given from there platform / version and extension are known)?

**Comment by begelundmuller** [comment]
Sorry, it's an open Discord, I thought it would automatically direct you through. The invite link is https://discord.gg/eEvSYHdfWK (you can also find it at the bottom of docs.rilldata.com).

Replicating the user's bug report here:
> Anyone else having this issue:
> > 2023-10-23T15:15:25.262    INFO    Rill shutdown gracefully
> > Error: reconcile project: failed to create catalog cache: HTTP Error: Failed to download extension "httpfs" at URL "http://extensions.duckdb.org/v0.9.1/osx_arm64/httpfs.duckdb_extension.gz"
> > Extension "httpfs" is an existing extension.
>
> I am on OSX, running either 0.36.1 or the nightly giving me the same error.

It was reported yesterday, and they are no longer seeing the issue today. So it seems this issue wasn't fully resolved last week.

**Comment by carlopi** [comment]
I believe the original issue was affecting a 12 hour span connected to the re-upload of the httpfs extension, and I think that has now be solved (as in assets are properly deployed and CDN has been flushed) and steps are being take to automatise the process of re-uploading extensions (and testing the process) so if / when this is required in the future this should be smooth.

---

As per punctual problems with the official extension repository, it's always very helpful if those are reported, and we might consider adding more context in the error messages and additional round of testing on our infrastructure.

If you are a user bumping against a similar problem, my advice would be to either download the extension via the secondary repository like:
```
INSTALL httpfs FROM 'http://nightly-extensions.duckdb.org';
```
Given this secondary repository has different settings, chances are one or the other will work.

If you are embedding DuckDB, there are a few more strategies that you could consider, one of them could be performing installation of core extensions directly as part of the installation process of DuckDB by issuing to DuckDB the relevant `INSTALL` instructions.

**Comment by carlopi** [comment]
> Maybe not it – but I looked briefly at the extension installation code and noticed it will fail on any HTTP status code that is not 200 ([source](https://github.com/duckdb/duckdb/blob/66be885660aab4d4bde16874ee6be1ed447bc19d/src/main/extension/extension_install.cpp#L281)). Could there be cases where your CDN succeeds with another status (2xx or 3xx)?

@begelundmuller: We will have to review CDN logs and perform more testing, there are some ideas for refactoring that area of the code to reuse httplib logic, that might bring also more robustness.

**Comment by mustafahasankhan** [comment]
@carlopi Seeing this issue yet again.

**Comment by carlopi** [comment]
@mustafahasankhan (or anyone else), can you share any more details (platform / geographic area) ?

I could not reproduce (yet), unsure what information would be useful.
Could you try like:
```
wget --tries=1 --verbose http://extensions.duckdb.org/022f01622/osx_amd64/fts.duckdb_extension.gz
```
or
```
wget --tries=1 --verbose http://extensions.duckdb.org/v0.9.1/osx_amd64/httpfs.duckdb_extension.gz
```

**Comment by dioptre** [comment]
Still a problem with http://extensions.duckdb.org/v0.9.1/linux_arm64_gcc4/httpfs.duckdb_extension.gz

**Comment by medriscoll** [comment]
We are seeing this issue again today, getting a 403 when attempting to pull the extension through our install scripts.

https://extensions.duckdb.org/v0.9.1/osx_arm64/httpsfs.duckdb_extension.gz

![image](https://github.com/duckdb/duckdb/assets/306103/49f09f23-d1ae-490e-93b9-6666db7f892f)

**Comment by carlopi** [comment]
> We are seeing this issue again today, getting a 403 when attempting to pull the extension through our install scripts.

Can you try against: https://extensions.duckdb.org/v0.9.1/osx_arm64/httpfs.duckdb_extension.gz
Unsure how you got to `httpsfs`, but the name of the extension should be `httpfs` (6 chars)

**Comment by begelundmuller** [comment]
@carlopi That seems like a typo from a reproduction attempt. Here is the exact error reported by a user on 10 April around 5pm CET:
```
HTTP Error: Failed to download extension "json" at URL "http://nightly-extensions.duckdb.org/v0.9.2/osx_arm64/json.duckdb_extension.gz" Extension "json" is an existing extension.
```

**Comment by begelundmuller** [comment]
I'm wondering if this error might be related to the user's network instead of the CDN, and the real network error message is being overridden by the `Extension "json" is an existing extension.` message?

**Comment by carlopi** [comment]
Error message should be additive, no information is deleted by the `Extension "json" is an existing extension` part, it's optional information concatenated to whatever the original error, and I checked and seemed to have been a very similar code also in v0.9.2.

I would possibly recommend a different approach for retries, where you could get the explicit path from DuckDB (even just parsing the error), then fetching the file with other means, potentially with higher timeouts / retry count, then pointing duckdb to INSTALL the local file you just fetched from the remote location.

That way you would both have more control over the response error and how to logging it, you can even perform more general testing of network, it would be valuable also to us and a feature we should consider adding, but while the general problem is a bit ill-defined, in your case you might restrict and decide what metrics are relevant to you, and your setup you are already looked to be close to this.

In general on transient failures of our S3-backed extension-serving infrastructure, it's hard to be sure, we can discuss this more in depth as part of the support offering you have in place, especially if you can find more evidence of potential problems.
